README.md
md5deep
=======

CMake buildsystem fork of Jesse Kornblum's md5deep/hashdeep.

This codebase is largely the original from SourceForge's SVN repo for the same
but has been pulled to here as a fork to make life easier for someone trying
to use it for cross-compilation and the like.  The Autotools build system
was broken in several ways that it was not worth trying to fix it properly 
for cross-compilation- so we have the start of the fork as we've moved it 
to CMake.


NEWS
** Changes in version 4.4beta1 (12-08-13)

* Removal of Autotools for build.
    
      Version of build system couldn't cross-compile, etc.  
      so it was re-worked to support CMake.  There's still
      liable to be issues on some officially supported 
      targets, but we're going to nab the stuff in the
      upstream SVN and ensure that at least Linux/*BSD/
      Windows works on build and operation.

** Changes in version 4.3 (24 Oct 2012)

* Bug Fixes

      Fixed check for endian-ness, affecting hash generation on
      big-endian platforms.

      Fixed minor bugs related to OpenSolaris.



** Changes in version 4.2 (11 Jun 2012)

* Bug Fixes

      Fixed padding in Tiger hashes for large files



** Changes in version 4.1.1 (5 Jun 2012)

* Bug Fixes

      Fixed hashdeep bug for known files with commas in their names.
      Cleaned up usage message


** Changes in version 4.1.0 (14 Feb 2012)

* New Features

      Added expert mode option to parse Windows PE files

* Bug Fixes

      Fixed junction point handling on Win32



** Changes in version 4.0.1 (22 Jan 2012)

* Bug Fixes

        Fixed hang on DFXML generation on Win32
	Fixed incorrect hashes via stdin on Win32
	Fixed "Too many open files" error on OS X
	Doc files in Win32 have been corrected. 



** Changes in version 4.0 (6 Jan 2012)

* New Features

      	Rewrote most of the program in C++.

	Enabled multiprocessor support on all platforms.

	Removed ten character limit on file size mode

* Bug Fixes

	Too numerous to count!



** Changes in version 3.9.1 (31 May 2011)

* Bug Fixes

      Fixed a cosmetic bugs regarding processing standard input
      in piecewise mode.

      Fixed a bug regarding processing known hash files which have
      parenthesis in the filename.

      Clarified documentation regarding multipliers in piecewise
      mode. They are IEC multipliers, or blocks of 1,024. Side note:
      also spelled 'multipliers' correctly.



** Changes in version 3.9 (21 Apr 2011)

* Bug Fixes

      Fixed a critical bug in hashdeep regarding processing
      standard input



** Changes in version 3.8 (15 Apr 2011)

* New Features

      Added -f mode to get the list of input files from
      a file on the disk.

      Added 64-bit executables for Windows.


* Bug Fixes

      Fixed a cosmetic bug regarding the error messages for 
      piecewise mode.




** Changes in version 3.7 (18 Dec 2010)

* New Features

      Added triage mode, -Z flag, to compute file size and partial
      hashes in addition to full file hash.

      Updated Windows cross-compiler to MinGW-4.2.1


* Bug Fixes
 
      Fixed hashdeep piecewise mode to include the correct size
      of each piece being hashed. This enables audit mode to 
      work correctly.

      Addressed display problem for time estimates of files with 
      long filenames in the current directory.

      Added "Input/Output error" to the list of fatal errors when
      processing files




** Changes in version 3.6 (23 Mar 2010)

* New Features

      Added expert mode found in md5deep to hashdeep.

* Bug Fixes

      Fixed error checking in helper routines.




** Changes in version 3.5.1 (30 Dec 2009)

* Bug Fixes

      Fixed processing of zero byte files.




** Changes in version 3.5 (22 Dec 2009)

* New Features

      Added mode to load list of files to process from a file (-f mode)

* Bug Fixes

      Fixed file size computation for piecewise hashing of standard input.

      Fixed error when auditing a directory to check for identical copies
      of a file if one copy is removed.



** Changes in version 3.4 (11 Jun 2009)

* Bug Fixes

      Fixed directory handle leak during recursive parsing on Windows

      Fixed -n mode (display unused hashes)



** Changes in version 3.3 (4 Apr 2009)

* New Features

* Bug Fixes

      Skip all reparse points (symbolic links and junction points)
      on Windows. Required for sane parsing of Windows Vista or 7 systems.

      Fixed handling of 64-bit timestamps on Windows.




** Changes in version 3.2 (25 Jan 2009)

* New Features

	Added Comma Separated Value (CSV) mode.


* Bug Fixes

	Fixed support for 64-bit files on Windows

	Fixed hashdeep input file parsing to allow filenames with 
	commas and properly handle other kinds of errors. Also 
	modified the invocation listing in output files such that
        even long lines can be successfully parsed.




** Changes in version 3.1 (23 Jul 2008)

* New Features

	Added size threshold modes to only process files smaller
	than a given size.

	Added a timestamp mode that records the creation time time
	for each file on Win32, the change time on all other operating
	systems.

	Streamlined file size computation process, which makes
	the programs about 15% faster.


* Bug Fixes

	Corrected time estimates for large files (e.g. files which 
        require more than one day).




** Changes in version 3.0 (10 May 2008)

* New Features

	Added hashdeep program to support multihashing and
        hash file auditing

	Added support for new iLook style hashes

* Bug Fixes
	
	Fixed obscure bug that caused a crash (double free) when
	attempting to check a very small file for EnCase hashes



** Changes in version 2.0.1 (31 Oct 2007)

* Bug Fixes

	Added support for older Linux systems that don't have the
	latest and greatest ioctl functions.



** Changes in version 2.0 (15 Oct 2007)

* New Features

	Using GNU Autotools for configuration and compilation. This
	should help avoid platform specific issues such as SHA-1
	problems previously found on 64-bit versions of AIX.

	Added support for files with Unicode characters in their
	filenames on Microsoft Windows.

	Added support for EnCase hash sets (.hash files).

	Updated web site and quick start guide

	Slightly reduced the size of all of the executables by removing
	duplicated code. 

* Bug Fixes

	Fixed time estimation mode for block devices on OS X and Linux


src/sha256.c
/*
 *  FIPS-180-2 compliant SHA-256 implementation
 *  written by Christophe Devine
 *
 *  This code has been distributed as PUBLIC DOMAIN.
 *
 *  Although normally licensed under the GPL on the author's web site,
 *  he has given me permission to distribute it as public domain as 
 *  part of md5deep. THANK YOU! Software authors are encouraged to
 *  use the GPL'ed version of this code available at:
 *  http://www.cr0.net:8040/code/crypto/sha256/ whenever possible.
 */

/* $Id: sha256.c 501 2011-09-25 17:12:27Z xchatty $ */

#include <string.h>
#include "sha256.h"


void hash_init_sha256(void * ctx)
{
    sha256_starts((context_sha256_t *)ctx);
}

void hash_update_sha256(void * ctx, const unsigned char *buf, size_t len)
{
  sha256_update((context_sha256_t *)ctx,buf,(uint32_t) len);
}

void hash_final_sha256(void * ctx, unsigned char *digest)
{
  sha256_finish((context_sha256_t *)ctx, digest);
}



#define GET_UINT32(n,b,i)                       \
{                                               \
    (n) = ( (uint32_t) (b)[(i)    ] << 24 )       \
        | ( (uint32_t) (b)[(i) + 1] << 16 )       \
        | ( (uint32_t) (b)[(i) + 2] <<  8 )       \
        | ( (uint32_t) (b)[(i) + 3]       );      \
}

#define PUT_UINT32(n,b,i)                       \
{                                               \
    (b)[(i)    ] = (uint8_t) ( (n) >> 24 );       \
    (b)[(i) + 1] = (uint8_t) ( (n) >> 16 );       \
    (b)[(i) + 2] = (uint8_t) ( (n) >>  8 );       \
    (b)[(i) + 3] = (uint8_t) ( (n)       );       \
}

void sha256_starts( context_sha256_t *ctx )
{
  ctx->total[0] = 0;
  ctx->total[1] = 0;

  ctx->state[0] = 0x6A09E667;
  ctx->state[1] = 0xBB67AE85;
  ctx->state[2] = 0x3C6EF372;
  ctx->state[3] = 0xA54FF53A;
  ctx->state[4] = 0x510E527F;
  ctx->state[5] = 0x9B05688C;
  ctx->state[6] = 0x1F83D9AB;
  ctx->state[7] = 0x5BE0CD19;
}

void sha256_process( context_sha256_t *ctx, const uint8_t data[64] );
void sha256_process( context_sha256_t *ctx, const uint8_t data[64] )
{
  uint32_t temp1, temp2, W[64];
  uint32_t A, B, C, D, E, F, G, H;

  GET_UINT32( W[0],  data,  0 );
  GET_UINT32( W[1],  data,  4 );
  GET_UINT32( W[2],  data,  8 );
  GET_UINT32( W[3],  data, 12 );
  GET_UINT32( W[4],  data, 16 );
  GET_UINT32( W[5],  data, 20 );
  GET_UINT32( W[6],  data, 24 );
  GET_UINT32( W[7],  data, 28 );
  GET_UINT32( W[8],  data, 32 );
  GET_UINT32( W[9],  data, 36 );
  GET_UINT32( W[10], data, 40 );
  GET_UINT32( W[11], data, 44 );
  GET_UINT32( W[12], data, 48 );
  GET_UINT32( W[13], data, 52 );
  GET_UINT32( W[14], data, 56 );
  GET_UINT32( W[15], data, 60 );

#define  SHR(x,n) ((x & 0xFFFFFFFF) >> n)
#define ROTR(x,n) (SHR(x,n) | (x << (32 - n)))

#define S0(x) (ROTR(x, 7) ^ ROTR(x,18) ^  SHR(x, 3))
#define S1(x) (ROTR(x,17) ^ ROTR(x,19) ^  SHR(x,10))

#define S2(x) (ROTR(x, 2) ^ ROTR(x,13) ^ ROTR(x,22))
#define S3(x) (ROTR(x, 6) ^ ROTR(x,11) ^ ROTR(x,25))

#define F0(x,y,z) ((x & y) | (z & (x | y)))
#define F1(x,y,z) (z ^ (x & (y ^ z)))

#define R(t)                                    \
(                                               \
    W[t] = S1(W[t -  2]) + W[t -  7] +          \
           S0(W[t - 15]) + W[t - 16]            \
)

#define P(a,b,c,d,e,f,g,h,x,K)                  \
{                                               \
    temp1 = h + S3(e) + F1(e,f,g) + K + x;      \
    temp2 = S2(a) + F0(a,b,c);                  \
    d += temp1; h = temp1 + temp2;              \
}

  A = ctx->state[0];
  B = ctx->state[1];
  C = ctx->state[2];
  D = ctx->state[3];
  E = ctx->state[4];
  F = ctx->state[5];
  G = ctx->state[6];
  H = ctx->state[7];

  P( A, B, C, D, E, F, G, H, W[ 0], 0x428A2F98 );
  P( H, A, B, C, D, E, F, G, W[ 1], 0x71374491 );
  P( G, H, A, B, C, D, E, F, W[ 2], 0xB5C0FBCF );
  P( F, G, H, A, B, C, D, E, W[ 3], 0xE9B5DBA5 );
  P( E, F, G, H, A, B, C, D, W[ 4], 0x3956C25B );
  P( D, E, F, G, H, A, B, C, W[ 5], 0x59F111F1 );
  P( C, D, E, F, G, H, A, B, W[ 6], 0x923F82A4 );
  P( B, C, D, E, F, G, H, A, W[ 7], 0xAB1C5ED5 );
  P( A, B, C, D, E, F, G, H, W[ 8], 0xD807AA98 );
  P( H, A, B, C, D, E, F, G, W[ 9], 0x12835B01 );
  P( G, H, A, B, C, D, E, F, W[10], 0x243185BE );
  P( F, G, H, A, B, C, D, E, W[11], 0x550C7DC3 );
  P( E, F, G, H, A, B, C, D, W[12], 0x72BE5D74 );
  P( D, E, F, G, H, A, B, C, W[13], 0x80DEB1FE );
  P( C, D, E, F, G, H, A, B, W[14], 0x9BDC06A7 );
  P( B, C, D, E, F, G, H, A, W[15], 0xC19BF174 );
  P( A, B, C, D, E, F, G, H, R(16), 0xE49B69C1 );
  P( H, A, B, C, D, E, F, G, R(17), 0xEFBE4786 );
  P( G, H, A, B, C, D, E, F, R(18), 0x0FC19DC6 );
  P( F, G, H, A, B, C, D, E, R(19), 0x240CA1CC );
  P( E, F, G, H, A, B, C, D, R(20), 0x2DE92C6F );
  P( D, E, F, G, H, A, B, C, R(21), 0x4A7484AA );
  P( C, D, E, F, G, H, A, B, R(22), 0x5CB0A9DC );
  P( B, C, D, E, F, G, H, A, R(23), 0x76F988DA );
  P( A, B, C, D, E, F, G, H, R(24), 0x983E5152 );
  P( H, A, B, C, D, E, F, G, R(25), 0xA831C66D );
  P( G, H, A, B, C, D, E, F, R(26), 0xB00327C8 );
  P( F, G, H, A, B, C, D, E, R(27), 0xBF597FC7 );
  P( E, F, G, H, A, B, C, D, R(28), 0xC6E00BF3 );
  P( D, E, F, G, H, A, B, C, R(29), 0xD5A79147 );
  P( C, D, E, F, G, H, A, B, R(30), 0x06CA6351 );
  P( B, C, D, E, F, G, H, A, R(31), 0x14292967 );
  P( A, B, C, D, E, F, G, H, R(32), 0x27B70A85 );
  P( H, A, B, C, D, E, F, G, R(33), 0x2E1B2138 );
  P( G, H, A, B, C, D, E, F, R(34), 0x4D2C6DFC );
  P( F, G, H, A, B, C, D, E, R(35), 0x53380D13 );
  P( E, F, G, H, A, B, C, D, R(36), 0x650A7354 );
  P( D, E, F, G, H, A, B, C, R(37), 0x766A0ABB );
  P( C, D, E, F, G, H, A, B, R(38), 0x81C2C92E );
  P( B, C, D, E, F, G, H, A, R(39), 0x92722C85 );
  P( A, B, C, D, E, F, G, H, R(40), 0xA2BFE8A1 );
  P( H, A, B, C, D, E, F, G, R(41), 0xA81A664B );
  P( G, H, A, B, C, D, E, F, R(42), 0xC24B8B70 );
  P( F, G, H, A, B, C, D, E, R(43), 0xC76C51A3 );
  P( E, F, G, H, A, B, C, D, R(44), 0xD192E819 );
  P( D, E, F, G, H, A, B, C, R(45), 0xD6990624 );
  P( C, D, E, F, G, H, A, B, R(46), 0xF40E3585 );
  P( B, C, D, E, F, G, H, A, R(47), 0x106AA070 );
  P( A, B, C, D, E, F, G, H, R(48), 0x19A4C116 );
  P( H, A, B, C, D, E, F, G, R(49), 0x1E376C08 );
  P( G, H, A, B, C, D, E, F, R(50), 0x2748774C );
  P( F, G, H, A, B, C, D, E, R(51), 0x34B0BCB5 );
  P( E, F, G, H, A, B, C, D, R(52), 0x391C0CB3 );
  P( D, E, F, G, H, A, B, C, R(53), 0x4ED8AA4A );
  P( C, D, E, F, G, H, A, B, R(54), 0x5B9CCA4F );
  P( B, C, D, E, F, G, H, A, R(55), 0x682E6FF3 );
  P( A, B, C, D, E, F, G, H, R(56), 0x748F82EE );
  P( H, A, B, C, D, E, F, G, R(57), 0x78A5636F );
  P( G, H, A, B, C, D, E, F, R(58), 0x84C87814 );
  P( F, G, H, A, B, C, D, E, R(59), 0x8CC70208 );
  P( E, F, G, H, A, B, C, D, R(60), 0x90BEFFFA );
  P( D, E, F, G, H, A, B, C, R(61), 0xA4506CEB );
  P( C, D, E, F, G, H, A, B, R(62), 0xBEF9A3F7 );
  P( B, C, D, E, F, G, H, A, R(63), 0xC67178F2 );

  ctx->state[0] += A;
  ctx->state[1] += B;
  ctx->state[2] += C;
  ctx->state[3] += D;
  ctx->state[4] += E;
  ctx->state[5] += F;
  ctx->state[6] += G;
  ctx->state[7] += H;
}

void sha256_update( context_sha256_t *ctx, const uint8_t *input, uint32_t length )
{
  uint32_t left, fill;

  if( ! length ) return;

  left = ctx->total[0] & 0x3F;
  fill = 64 - left;

  ctx->total[0] += length;
  ctx->total[0] &= 0xFFFFFFFF;

  if( ctx->total[0] < length )
    ctx->total[1]++;

  if( left && length >= fill )
    {
      memcpy( (void *) (ctx->buffer + left),
	      (const void *) input, fill );
      sha256_process( ctx, ctx->buffer );
      length -= fill;
      input  += fill;
      left = 0;
    }

  while( length >= 64 )
    {
      sha256_process( ctx, input );
      length -= 64;
      input  += 64;
    }

  if( length )
    {
      memcpy( (void *) (ctx->buffer + left),
	      (const void *) input, length );
    }
}

static uint8_t sha256_padding[64] =
  {
    0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
  };

void sha256_finish( context_sha256_t *ctx, uint8_t digest[32] )
{
  uint32_t last, padn;
  uint32_t high, low;
  uint8_t msglen[8];

  high = ( ctx->total[0] >> 29 )
    | ( ctx->total[1] <<  3 );
  low  = ( ctx->total[0] <<  3 );

  PUT_UINT32( high, msglen, 0 );
  PUT_UINT32( low,  msglen, 4 );

  last = ctx->total[0] & 0x3F;
  padn = ( last < 56 ) ? ( 56 - last ) : ( 120 - last );

  sha256_update( ctx, sha256_padding, padn );
  sha256_update( ctx, msglen, 8 );

  PUT_UINT32( ctx->state[0], digest,  0 );
  PUT_UINT32( ctx->state[1], digest,  4 );
  PUT_UINT32( ctx->state[2], digest,  8 );
  PUT_UINT32( ctx->state[3], digest, 12 );
  PUT_UINT32( ctx->state[4], digest, 16 );
  PUT_UINT32( ctx->state[5], digest, 20 );
  PUT_UINT32( ctx->state[6], digest, 24 );
  PUT_UINT32( ctx->state[7], digest, 28 );
}

#ifdef TEST

#include <stdlib.h>
#include <stdio.h>

/*
 * those are the standard FIPS-180-2 test vectors
 */

static char *msg[] = 
  {
    "abc",
    "abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq",
    NULL
  };

static char *val[] =
  {
    "ba7816bf8f01cfea414140de5dae2223" \
    "b00361a396177a9cb410ff61f20015ad",
    "248d6a61d20638b8e5c026930c3e6039" \
    "a33ce45964ff2167f6ecedd419db06c1",
    "cdc76e5c9914fb9281a1c7e284d73e67" \
    "f1809a48a497200e046d39ccc7112cd0"
  };

int main( int argc, char *argv[] )
{
  FILE *f;
  int i, j;
  char output[65];
  context_sha256_t ctx;
  unsigned char buf[1000];
  unsigned char sha256sum[32];

  if( argc < 2 )
    {
      printf( "\n SHA-256 Validation Tests:\n\n" );

      for( i = 0; i < 3; i++ )
        {
	  printf( " Test %d ", i + 1 );

	  sha256_starts( &ctx );

	  if( i < 2 )
            {
	      sha256_update( &ctx, (uint8_t *) msg[i],
			     strlen( msg[i] ) );
            }
	  else
            {
	      memset( buf, 'a', 1000 );

	      for( j = 0; j < 1000; j++ )
                {
		  sha256_update( &ctx, (uint8_t *) buf, 1000 );
                }
            }

	  sha256_finish( &ctx, sha256sum );

	  for( j = 0; j < 32; j++ )
            {
	      sprintf( output + j * 2, "%02x", sha256sum[j] );
            }

	  if( memcmp( output, val[i], 64 ) )
            {
	      printf( "failed!\n" );
	      return( 1 );
            }

	  printf( "passed.\n" );
        }

      printf( "\n" );
    }
  else
    {
      if( ! ( f = fopen( argv[1], "rb" ) ) )
        {
	  perror( "fopen" );
	  return( 1 );
        }

      sha256_starts( &ctx );

      while( ( i = fread( buf, 1, sizeof( buf ), f ) ) > 0 )
        {
	  sha256_update( &ctx, buf, i );
        }

      sha256_finish( &ctx, sha256sum );

      for( j = 0; j < 32; j++ )
        {
	  printf( "%02x", sha256sum[j] );
        }

      printf( "  %s\n", argv[1] );
    }

  return( 0 );
}

#endif
